Refactor Feedforward Angle and RC Smoothing - mashup of 12578 and 12594

ctzsnooze opened the pull request:
<Ready to merge now, I think>

@haslinghuis I've allowed feedforward max rate limit above 150, now up to 200, which is not likely to cause any issues and I doubt many people will want to use above 150, but perhaps at some later point we can modify configurator.

This PR combines #12578 and #12594, which are now closed.  Both shared a lot of common changes in pid.c and rc.c, and would be difficult to merge sequentially.  It includes the current updated Angle and Horizon code.

There are two functional changes:

1.  Delayed onset of levelling in Horizon when returning sticks to the centre.  As suggested by @JustainChoe, this adds a delay whenever the levelling strength increases in Horizon mode.  Decreases in levelling strength are unaffected and happen immediately.  The delay is achieved by applying a first order lowpass filter, with a time constant of 500ms by default to the horizon strength factor, where a `horizon_delay` of 50 means 500ms.  The code then uses the lowest of the un-smoothed and smoothed value, where a low value means no smoothing.  This results in a delay only when the horizon strength value increases.  The pilot can do a flip and bring their sticks hard back to centre, and the levelling will come on relatively slowly, avoiding the previously unwelcome 'snap' due to the combination of self-levelling and the pilot bringing the sticks to a halt.  It there's too much 'snap', or if you want the levelling to come in only after a longer time, increase the time constant (up to 250 or 2.5s).  For more immediate levelling, as perhaps preferred by a person learning Acro via Horizon, the delay can be reduced.  If set to zero, there is no delay.

2.  Small modification to `feedforward_max_rate_limit`.  Now this value affects the slope of the limiting line, with zero feedforward always by the time setpoint reaches maximum, rather than moving the line left and right, while the users PID "P" value modifies the slow.  Racers may prefer `feedforward_max_rate_limit` at values around 100-120 to maintain full feedforward until quite late in a full stick move.  Cinematic users or those with heavier quads may want a smoother, less aggressive feedforward, in which case `feedforward_max_rate_limit` could be around 50, and the RC smoothing on feedforward increased a bit (lower cutoff than default).

Please refer to #12576 and #12594 for detailed notes on the other changes..

There is a lot of efficiency related refactoring, resulting in significant CPU and flash space savings.

Summary of changes:
- bug fix for an issue that caused feedforward RC smoothing to get a different value from setpoint in auto-rc-smoothing mode
- fixes issues that caused the RC link speed to be calculated incorrectly or inaccurately, especially at high telemetry rates.  - rc_smoothing calculations now use lowpass filtering with much simpler code to detect link speed changes - many thanks to @KarateBrot for advice on how to best do this.  RC smoothing is now updated every 3 valid samples.
- fixes an issue where the user's PID P value altered the feedforward max_rate_limit_factor attenuation slope.  This was by design, but a bit annoying, and it was confusing that high P caused feedforward to be stronger at the start of flips, and vice versa.  Now the feedforward_max_rate_limit value controls the slope of the attenuation line, so that when it is set low, feedforward will, at the start of a flip, ease back earlier, and when set high, stay strong for later.  Default of 90 remains the same and suits high authority quads.  Quads that overshoot at the start of a flip should consider a lower than default feedforward_max_rate_limit.
- fixes and issue where Horizon mode would get RC stepping in the levelling element with slow link speeds
- user-adjustment of Angle Mode Earth Referencing strength is now possible via OSD
- refactoring and simplification of the feedforward code to improve efficiency (many changes)
- minor fix to stop overflow of debug values in rc_smoothing_rate debug
- quite a few debug changes since some parameters don't exist anymore
- editorial changes to rename variables named 'rate' when the value was a time interval
- about 1k less flash space required
- On F411, 3% less CPU in acro, 4-4.5% less in angle/horizon.

There should be no change in feedforward or RC smoothing functionality.  All quads should fly the same, just a bit more smoothly.  CPU load / instability should be a bit less of a problem on marginal setups.

RC smoothing values and reported link speeds should be accurate in all circumstances.  Previously when the RC smoothing calculations were in error, RC link stepping could get through to the motors via feedforward or setpoint.  This shouldn't happen any more, including in Angle and Horizon mode.

The code should be able to deal with radios that change links dynamically, even in small steps.  Previously there was an issue detecting change from 333 to 250hz.  Now it should respond to any change greater than 1hz, and do so within a few packets.

Please also test feedforward by flying and noting normal feedforward behaviour, and log feedforward.

If you find any issues with unexpected link intervals or rc smoothing issues, please either raise an issue here on GitHub or a pre-issue on Discord.

Feedforward, when a new RC packet arrives, requires 125 cycles without averaging, jitter or transition.  This increases to 170 cycles with averaging, 196 cycles when also adding jitter reduction, and 209 cycles with "the lot". 

RC Smoothing requires 18 cycles while idle, about 160 cycles when a new RC packet arrives in that is not much different from the previous interval, and about 250 cycles when the interval is sufficiently different that new cutoffs are required.

Recoding for efficiency mostly involved:
- not using "#DEFINE" values, instead using constants in the code, or setting the value when declaring the variable that needed it, with a comment
- avoiding unnecessary int to float and float to int conversions, especially 'hidden' or inside loops
- generally using floats if at any point they would be multiplied by a float or divided by a float
- using const variables whenever possible, especially for values that were repeatedly calculated inside loops
- pre-calculating values required for divisions in the pidRuntime init, so we only need to do multiplications.

blckmn commented:
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> FAIL
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

nerdCopter commented:
have no blackbox on this, but .. if it helps:
```
# rc_smoothing_info
# RC Smoothing Type: FILTER
# Detected RX frame rate: 2.003ms
# Active setpoint cutoff: 58hz (auto)
# Active FF cutoff: 58hz (auto)
# Active throttle cutoff: 187hz (auto)
```
TX setup for 500hz, 1:4 telem

nerdCopter commented:
`e0d143819` displayed `RC_SMOOTHING` warning entire flight.
```
# 

# version
# Betaflight / STM32F411 (S411) 4.5.0 Apr  4 2023 / 22:11:47 (e0d143819) MSP API: 1.46
# board: manufacturer_id: BEFH, board_name: BETAFPVF4SX1280

# status
MCU F411 Clock=108MHz (PLLP-HSE), Vref=3.27V, Core temp=45degC
Stack size: 2048, Stack address: 0x2001fff0
Configuration: CONFIGURED, size: 3511, max available: 16384
Devices detected: SPI:1, I2C:0
Gyros detected: gyro 1 locked dma
GYRO=ICM20689, ACC=ICM20689
OSD: MAX7456 (30 x 13)
BUILD KEY: 519948a11ec5c95f660a88281d8d19de (4.5.0-zulu)
System Uptime: 35 seconds, Current Time: 2023-04-04T22:35:15.485+00:00
CPU:24%, cycle time: 124, GYRO rate: 8064, RX rate: 15, System rate: 9
Voltage: 8 * 0.01V (0S battery - NOT PRESENT)
I2C Errors: 0
Arming disable flags: RXLOSS CLI MSP

# tasks
Task list             rate/hz  max/us  avg/us maxload avgload  total/ms   late    run reqd/us
00 - (         SYSTEM)     10       3       1    0.0%    0.0%         0      0    479       1
01 - (         SYSTEM)    990       9       0    0.8%    0.0%        30    246  47038       4
02 - (           GYRO)   8006      11       3    8.8%    2.9%      1322      0 381448       0
03 - (         FILTER)   4005      13       4    5.2%    1.8%       902      0 190724       0
04 - (            PID)   4005      66      45   26.4%   18.3%      8656      0 190724       0
05 - (            ACC)    991      14       3    1.3%    0.3%       149     48  46992       9
06 - (       ATTITUDE)     99      21      11    0.2%    0.1%        51      4   4759      20
07 - (             RX)     10      33      19    0.0%    0.0%        28     11   1437      22
08 - (         SERIAL)     97  478232       1 4638.8%    0.0%       847      0   4726      61
09 - (       DISPATCH)    989       8       0    0.7%    0.0%        38    157  47046       5
10 - (BATTERY_VOLTAGE)    199      11       2    0.2%    0.0%        19     15   9508       7
12 - ( BATTERY_ALERTS)      5       4       2    0.0%    0.0%         0      1    241       3
13 - (         BEEPER)     97       9       0    0.0%    0.0%         4      7   4726       3
14 - (      TELEMETRY)    495      27       0    1.3%    0.0%        18     85  23518       2
15 - (       LEDSTRIP)     99     168       7    1.6%    0.0%        42      1   4703      63
16 - (            OSD)     12      96      21    0.1%    0.0%       339     63  18676      29
17 - (     ESC_SENSOR)     98       8       1    0.0%    0.0%         5     11   4726       3
18 - (            CMS)     20       7       1    0.0%    0.0%         1      2    954       4
19 - (        VTXCTRL)      5      75       1    0.0%    0.0%         1      2    240      65
21 - (    ADCINTERNAL)      1       9       2    0.0%    0.0%         0      1     49       8
22 - (SPEED_NEGOTIATION)     98      12       2    0.1%    0.0%         6      9   4726      10
RX Check Function                  17       6                       123
Total (excluding SERIAL)                                23.4%

# rc_smoothing_info
# RC Smoothing Type: FILTER
# Detected RX frame rate: NO SIGNAL
# Active setpoint cutoff: 0hz (auto)
# Active FF cutoff: 0hz (auto)
# Active throttle cutoff: 0hz (auto)

```

TheIsotopes commented:
I can confirm, same problem as @nerdCopter ... seems like `RC_Smoothing` is not active.

[BTFL_BLACKBOX_LOG_Egodrift_20230404.txt](https://github.com/betaflight/betaflight/files/11155900/BTFL_BLACKBOX_LOG_Egodrift_20230404.txt)

nerdCopter commented:
no OSD warning this time. :)
```
# version
# Betaflight / STM32F411 (S411) 4.5.0 Apr  5 2023 / 22:12:30 (0ddf5925d) MSP API: 1.46
# board: manufacturer_id: BEFH, board_name: BETAFPVF4SX1280

# rc_smoothing_info
# RC Smoothing Type: FILTER
# Detected RX frame rate: 2.004ms
# Active setpoint cutoff: 58hz (auto)
# Active FF cutoff: 58hz (auto)
# Active throttle cutoff: 187hz (auto)
```
SPI ELRS; TX 500hz, 1:4 telem

haslinghuis reviewed (dismissed):
Only needs to update unit tests and remove comments.

TheIsotopes commented:
tested again with rebased PR on ghost and frsky sbus.
everything flies as expected, to me all looks fine. 👍

nerdCopter commented:
```
# version
# Betaflight / STM32F411 (S411) 4.5.0 Apr  7 2023 / 17:27:59 (cd9c2e5ee) MSP API: 1.46
# board: manufacturer_id: BEFH, board_name: BETAFPVF4SX1280

# rc_smoothing_info
# RC Smoothing Type: FILTER
# Detected Rx frequency: 499Hz
# Active setpoint cutoff: 57hz (auto)
# Active FF cutoff: 57hz (auto)
# Active throttle cutoff: 187hz (auto)
```
SPI ELRS; TX 500hz, 1:4 telem

nerdCopter commented:
back yard hover test (6" w/ 5" props): [BTFL_BLACKBOX_LOG_APEX-6inch_20230407_165227_tuning_8.BBL.zip](https://github.com/betaflight/betaflight/files/11181749/BTFL_BLACKBOX_LOG_APEX-6inch_20230407_165227_tuning_8.BBL.zip)

ELRS; TX 500hz, 1:4 telem
```
# 

# version
# Betaflight / STM32F7X2 (S7X2) 4.5.0 Apr  7 2023 / 18:30:53 (cd9c2e5ee) MSP API: 1.46
# board: manufacturer_id: APEX, board_name: APEXF7

# rc_smoothing_info
# RC Smoothing Type: FILTER
# Detected Rx frequency: 500Hz
# Active setpoint cutoff: 25hz (manual)
# Active FF cutoff: 25hz (manual)
# Active throttle cutoff: 187hz (auto)

# status
MCU F722 Clock=216MHz, Vref=3.27V, Core temp=51degC
Stack size: 2048, Stack address: 0x20010000
Configuration: CONFIGURED, size: 3353, max available: 16384
Devices detected: SPI:1, I2C:1
Gyros detected: gyro 1 locked dma
GYRO=MPU6000, ACC=MPU6000, BARO=BMP280
OSD: MAX7456 (30 x 13)
BUILD KEY: da4db58bae5e46f744dac328f38988d0 (4.5.0-zulu)
System Uptime: 30 seconds, Current Time: 2023-04-07T21:58:18.200+00:00
CPU:32%, cycle time: 124, GYRO rate: 8064, RX rate: 500, System rate: 9
Voltage: 297 * 0.01V (0S battery - NOT PRESENT)
I2C Errors: 0
FLASH: JEDEC ID=0x00ef4018 16M
Arming disable flags: NOPREARM CLI MSP

# tasks
Task list             rate/hz  max/us  avg/us maxload avgload  total/ms   late    run reqd/us
00 - (         SYSTEM)     10       7       0    0.0%    0.0%         0      3    324       5
01 - (         SYSTEM)    990      13       0    1.2%    0.0%        20     55  31922       6
02 - (           GYRO)   8013      12       1    9.6%    1.4%       415      0 258399       0
03 - (         FILTER)   8013      26       9   20.8%    7.2%      2384      0 258399       0
04 - (            PID)   8013      54      24   43.2%   19.7%      6658      0 258399       0
05 - (            ACC)    997      17       3    1.6%    0.3%        97     33  31898      11
06 - (       ATTITUDE)    100      27       9    0.2%    0.0%        31      8   3215      25
07 - (             RX)    363      36      16    1.3%    0.5%       548     47  35058      20
08 - (         SERIAL)     99  294035       1 2910.9%    0.0%       527      0   3201      64
09 - (       DISPATCH)    998      13       0    1.2%    0.0%        21     49  31927       2
10 - (BATTERY_VOLTAGE)     50       9       1    0.0%    0.0%         3      3   1610       5
11 - (BATTERY_CURRENT)     50      14       1    0.0%    0.0%         1      1   1610       4
12 - ( BATTERY_ALERTS)      5       6       2    0.0%    0.0%         0      0    163       5
13 - (         BEEPER)     99      13       1    0.1%    0.0%         2      3   3200       4
14 - (           BARO)     40      44      23    0.1%    0.0%        88     10   3850      37
15 - (       ALTITUDE)     99      12       2    0.1%    0.0%         7      2   3200      10
16 - (      TELEMETRY)    496      14       0    0.6%    0.0%        16     26  15979       2
17 - (            OSD)     12      62      15    0.0%    0.0%       127     58   9588      31
19 - (            CMS)     20       8       1    0.0%    0.0%         1      2    646       7
20 - (        VTXCTRL)      5      33       1    0.0%    0.0%         1      1    163      31
21 - (        CAMCTRL)      5       4       0    0.0%    0.0%         0      3    163       3
23 - (    ADCINTERNAL)      1      10       3    0.0%    0.0%         0      2     33       9
24 - (       PINIOBOX)     20       9       2    0.0%    0.0%         1      1    646       8
25 - (SPEED_NEGOTIATION)     99      10       1    0.0%    0.0%         4      6   3200       3
RX Check Function                  13       1                        48
Total (excluding SERIAL)                                29.1%

# 
```

TheIsotopes commented:
Even after the last update of this PR everything looks great.
Everything flies as expected and the logs look the same as always.

nerdCopter commented:
@TheIsotopes , agreed, i don't fly angle/horizon either.  
and don't forget `set debug_mode = RC_SMOOTHING_RATE`.

TheIsotopes commented:
as before and also this time everything works without problems. 👍

[bbl_ghost_sbus.zip](https://github.com/betaflight/betaflight/files/11192003/bbl_ghost_sbus.zip)

nerdCopter commented:
back yard hover in stable-mode (kind of freaky and don't like testing that mode :laughing:). 
`set debug_mode = RC_SMOOTHING_RATE`
[BTFL_BLACKBOX_LOG_APEX-6inch_20230410_134142_stable_hover.BBL.zip](https://github.com/betaflight/betaflight/files/11193027/BTFL_BLACKBOX_LOG_APEX-6inch_20230410_134142_stable_hover.BBL.zip)

ELRS; TX 500hz, 1:4 telem
```
# 

# version
# Betaflight / STM32F7X2 (S7X2) 4.5.0 Apr 10 2023 / 14:11:27 (cd272ef1c) MSP API: 1.46
# board: manufacturer_id: APEX, board_name: APEXF7

# rc_smoothing_info
# RC Smoothing Type: FILTER
# Detected Rx frequency: 501Hz
# Active setpoint cutoff: 25hz (manual)
# Active FF cutoff: 25hz (manual)
# Active throttle cutoff: 187hz (auto)

# status
MCU F722 Clock=216MHz, Vref=3.27V, Core temp=50degC
Stack size: 2048, Stack address: 0x20010000
Configuration: CONFIGURED, size: 3353, max available: 16384
Devices detected: SPI:1, I2C:1
Gyros detected: gyro 1 locked dma
GYRO=MPU6000, ACC=MPU6000, BARO=BMP280
OSD: MAX7456 (30 x 13)
BUILD KEY: 46393bfedffac0532d561091cf5071c4 (4.5.0-zulu)
System Uptime: 158 seconds, Current Time: 2023-04-10T18:44:07.841+00:00
CPU:32%, cycle time: 124, GYRO rate: 8064, RX rate: 250, System rate: 9
Voltage: 305 * 0.01V (1S battery - CRITICAL)
I2C Errors: 0
FLASH: JEDEC ID=0x00ef4018 16M
Arming disable flags: NOPREARM CLI MSP

# tasks
Task list             rate/hz  max/us  avg/us maxload avgload  total/ms   late    run reqd/us
00 - (         SYSTEM)     10       6       0    0.0%    0.0%         1      3   1577       4
01 - (         SYSTEM)    997      14       0    1.3%    0.0%        99    232 145118       4
02 - (           GYRO)   8013      18       1   14.4%    1.2%      1923      0 1168261       0
03 - (         FILTER)   8013      29       9   23.2%    7.2%     10692      0 1168261       0
04 - (            PID)   8013      63      26   50.4%   20.8%     31102      0 1168261       0
05 - (            ACC)    997      18       2    1.7%    0.2%       403    144 145004       9
06 - (       ATTITUDE)    100      27       9    0.2%    0.0%       148     21  15006      14
07 - (             RX)    359      36      16    1.2%    0.6%      2669    215 164301      18
08 - (         SERIAL)    100  296866       1 2968.6%    0.0%     13895      0  14443      15
09 - (       DISPATCH)    995      14       0    1.3%    0.0%        95    229 145133       3
10 - (BATTERY_VOLTAGE)     50      16       2    0.0%    0.0%        16      6   7545       8
11 - (BATTERY_CURRENT)     50      17       2    0.0%    0.0%        18     11   7533       8
12 - ( BATTERY_ALERTS)      5      10       3    0.0%    0.0%         2      1    793       7
13 - (         BEEPER)    100      16       0    0.1%    0.0%        16     12  14233       7
14 - (           BARO)     40      44      23    0.1%    0.0%       428     13  18031      38
15 - (       ALTITUDE)    100      16       3    0.1%    0.0%        43     19  14233       9
16 - (      TELEMETRY)    499      23       0    1.1%    0.0%       152     74  71206       4
17 - (            OSD)     12      57      16    0.0%    0.0%       643     96  51765      31
19 - (            CMS)     20      13       2    0.0%    0.0%         6      5   3050       6
20 - (        VTXCTRL)      5      42       1    0.0%    0.0%        18      5    792      37
21 - (        CAMCTRL)      5       8       0    0.0%    0.0%         0      2    792       4
23 - (    ADCINTERNAL)      1      10       3    0.0%    0.0%         0      0    160       9
24 - (       PINIOBOX)     20      14       2    0.0%    0.0%         7      1   3046       7
25 - (SPEED_NEGOTIATION)    100      13       1    0.1%    0.0%        18     15  14233       6
RX Check Function                  15       0                       243
Total (excluding SERIAL)                                30.0%

# 
```

TheIsotopes commented:
@ctzsnooze pls can you rebase, for testing on latest master?

KarateBrot commented on src/main/flight/pid.c:
```diff
@@ -119,7 +118,7 @@ PG_RESET_TEMPLATE(pidConfig_t, pidConfig,
 #define LAUNCH_CONTROL_YAW_ITERM_LIMIT 50 // yaw iterm windup limit when launch mode is "FULL" (all axes)
 
-PG_REGISTER_ARRAY_WITH_RESET_FN(pidProfile_t, PID_PROFILE_COUNT, pidProfiles, PG_PID_PROFILE, 6);
+PG_REGISTER_ARRAY_WITH_RESET_FN(pidProfile_t, PID_PROFILE_COUNT, pidProfiles, PG_PID_PROFILE, 7);
```
I don't think we need to increment this. I don't see any changes in `pidProfile_t`.

KarateBrot commented on src/main/flight/pid.h:
```diff
@@ -456,16 +450,22 @@ void applyItermRelax(const int axis, const float iterm,
 float pidGetDT();
 float pidGetPidFrequency();
+
 float pidGetFeedforwardBoostFactor();
 float pidGetFeedforwardSmoothFactor();
 float pidGetFeedforwardJitterFactor();
+float pidGetFeedforwardAveraging();
 float pidGetFeedforwardTransitionFactor();
+float pidGetFeedforwardMaxRateLimit();
+float pidGetFeedforwardPidKp(int axis);
```
You can replace all these functions with this one:
```suggestion
pidRuntime_t *pidGetPidRuntime(void);
```
This comment makes more sense if you read the long comment in `/main/fc/rc.c`

ctzsnooze commented on src/main/cms/cms_menu_imu.c:
```diff
@@ -678,6 +678,7 @@ static const OSD_Entry cmsx_menuProfileOtherEntries[] = {
     { "ANGLE P",         OME_UINT8,  NULL, &(OSD_UINT8_t)  { &cmsx_angleP,                     0,    200,   1  }    },
     { "ANGLE FF",        OME_UINT8,  NULL, &(OSD_UINT8_t)  { &cmsx_angleFF,                    0,    200,   1  }    },
     { "ANGLE LIMIT",     OME_UINT8,  NULL, &(OSD_UINT8_t)  { &cmsx_angleLimit,                10,     90,   1  }    },
+    { "ANGLE E_REf",     OME_UINT8,  NULL, &(OSD_UINT8_t)  { &cmsx_angleEarthRef,              0,    100,   1  }    },
```
Thanks, done!

ctzsnooze commented on src/main/fc/rc.c:
```diff
@@ -57,57 +57,55 @@
 static applyRatesFn *applyRates;
-static uint16_t currentRxRefreshRate;
+
+static uint16_t currentRxIntervalUs; // packet interval un microseconds
+static float currentRxFrequencyHz; // packet interval un microseconds
```
Great recommendation, thank you.  Have implemented `RxRate` throughout.

ctzsnooze commented on src/main/fc/rc.c:
```diff
@@ -57,57 +57,55 @@
+    return feedforwardRaw[axis];
+#endif
+}
+#endif
```
Thanks, done.

ctzsnooze commented on src/main/fc/rc.c:
```diff
@@ -279,124 +262,99 @@ void updateRcRefreshRate(timeUs_t currentTimeUs)
+    currentRxIntervalUs = constrain(frameDeltaUs, RX_INTERVAL_MIN_US, RX_INTERVAL_MAX_US);
+    isRxIntervalValid = frameDeltaUs == currentRxIntervalUs;
+
+    currentRxFrequencyHz = 1e6 / currentRxIntervalUs; // cannot be zero due to preceding constraint
```
Great catch, thank you!

ctzsnooze commented on src/main/fc/rc.c:
```diff
@@ -514,48 +454,179 @@ static FAST_CODE void processRcSmoothingFilter(void)
 }
 #endif // USE_RC_SMOOTHING_FILTER
 
+FAST_CODE_NOINLINE void initAveraging (void)
```
Thank you, I was unsure about this.

ctzsnooze replied:
Done

ctzsnooze commented on src/main/fc/rc.c:
```diff
@@ -514,48 +454,179 @@ static FAST_CODE void processRcSmoothingFilter(void)
+    }
+}
+
+FAST_CODE_NOINLINE void calculateFeedforward (int axis)
```
OK

nerdCopter commented:
`e710f2a89` back yard hover in stable-mode
set debug_mode = RC_SMOOTHING_RATE
ELRS; TX 500hz, 1:4 telem
2/2 logs: [BTFL_BLACKBOX_LOG_APEX-6inch_20230415_112542_PR12605_e710f2a89_stable_hovers.BBL.zip](https://github.com/betaflight/betaflight/files/11239993/BTFL_BLACKBOX_LOG_APEX-6inch_20230415_112542_PR12605_e710f2a89_stable_hovers.BBL.zip)

```
# 

# version
# Betaflight / STM32F7X2 (S7X2) 4.5.0 Apr 15 2023 / 16:06:45 (e710f2a89) MSP API: 1.46
# board: manufacturer_id: APEX, board_name: APEXF7

# rc_smoothing_info
# RC Smoothing Type: FILTER
# Detected Rx frequency: 501Hz
# Active setpoint cutoff: 25hz (manual)
# Active FF cutoff: 25hz (manual)
# Active throttle cutoff: 187hz (auto)

# status
MCU F722 Clock=216MHz, Vref=3.26V, Core temp=60degC
Stack size: 2048, Stack address: 0x20010000
Configuration: CONFIGURED, size: 3353, max available: 16384
Devices detected: SPI:1, I2C:1
Gyros detected: gyro 1 locked dma
GYRO=MPU6000, ACC=MPU6000, BARO=BMP280
OSD: MAX7456 (30 x 13)
BUILD KEY: 65d2eb69e99fe7118c5abe589764b140 (4.5.0-zulu)
System Uptime: 180 seconds, Current Time: 2023-04-15T16:28:04.675+00:00
CPU:33%, cycle time: 125, GYRO rate: 8000, RX rate: 502, System rate: 9
Voltage: 306 * 0.01V (1S battery - CRITICAL)
I2C Errors: 0
FLASH: JEDEC ID=0x00ef4018 16M
Arming disable flags: NOPREARM CLI MSP

# tasks
Task list             rate/hz  max/us  avg/us maxload avgload  total/ms   late    run reqd/us
00 - (         SYSTEM)     10       9       1    0.0%    0.0%         1      3   1786       4
01 - (         SYSTEM)    994      14       0    1.3%    0.0%        99    205 151711       3
02 - (           GYRO)   8013      19       1   15.2%    1.3%      2245      0 1219599       0
03 - (         FILTER)   8013      30       9   24.0%    7.2%     11144      0 1219599       0
04 - (            PID)   8013     124      26   99.3%   20.8%     34441      0 1219599       0
05 - (            ACC)    995      19       3    1.8%    0.3%       472    166 151571       6
06 - (       ATTITUDE)    100      29      10    0.2%    0.1%       167     44  16524      19
07 - (             RX)    388      35      14    1.3%    0.5%      2709    254 173181      15
08 - (         SERIAL)     99  293922       2 2909.8%    0.0%     29320      0  15529      47
09 - (       DISPATCH)    997      13       0    1.2%    0.0%       106    213 151747       3
10 - (BATTERY_VOLTAGE)     50      14       3    0.0%    0.0%        19     11   8111       7
11 - (BATTERY_CURRENT)     50      18       3    0.0%    0.0%        21      7   8103      10
12 - ( BATTERY_ALERTS)      5       9       4    0.0%    0.0%         2      2    900       7
13 - (         BEEPER)     99      12       0    0.1%    0.0%        18      6  15034       7
14 - (           BARO)     40      44      23    0.1%    0.0%       461     13  19816      42
15 - (       ALTITUDE)     99      19       3    0.1%    0.0%        42     20  15034       5
16 - (      TELEMETRY)    497      23       0    1.1%    0.0%       165     83  74507       4
17 - (            OSD)     12      51      15    0.0%    0.0%       668    142  55745      33
19 - (            CMS)     20      14       2    0.0%    0.0%         6      3   3345       5
20 - (        VTXCTRL)      5      38       1    0.0%    0.0%        21      5    895      35
21 - (        CAMCTRL)      5       5       1    0.0%    0.0%         0      2    897       4
23 - (    ADCINTERNAL)      1      14       4    0.0%    0.0%         0      1    181      13
24 - (       PINIOBOX)     20      14       2    0.0%    0.0%         7      2   3344       5
25 - (SPEED_NEGOTIATION)     99      12       1    0.1%    0.0%        20     17  15033       3
RX Check Function                  17       1                       272
Total (excluding SERIAL)                                30.2%

# 
```

ctzsnooze commented on src/main/cli/cli.c:
```diff
@@ -4845,12 +4845,13 @@ static void cliRcSmoothing(const char *cmdName, char *cmdline)
-            cliPrint("# Detected RX frame rate: ");
-            if (avgRxFrameUs == 0) {
+            const uint16_t smoothedRxFrequency = lrintf(rcSmoothingData->smoothedRxFrequencyHz);
+            int rxInterval = getCurrentRxIntervalUs();
```
Done!

ctzsnooze commented on src/main/cli/cli.c:
```diff
@@ -4845,12 +4845,13 @@ static void cliRcSmoothing(const char *cmdName, char *cmdline)
+            const uint16_t smoothedRxFrequency = lrintf(rcSmoothingData->smoothedRxFrequencyHz);
+            int rxInterval = getCurrentRxIntervalUs();
+            cliPrint("# Detected Rx frequency: ");
+            if (rxInterval == 0) {
```
Done

ctzsnooze commented on src/main/fc/rc.c:
```diff
@@ -514,48 +454,179 @@ static FAST_CODE void processRcSmoothingFilter(void)
+    const float feedforwardJitterFactor = pidGetFeedforwardJitterFactor();
+    const float feedforwardTransitionFactor = pidGetFeedforwardTransitionFactor();
+    const float feedforwardBoostFactor = pidGetFeedforwardBoostFactor();
+    const float maxRateLimit = pidGetFeedforwardMaxRateLimit();
```
Thank you very much for explaining how to do this.  Implemented the first method.  Works great!

ctzsnooze commented:
Thanks @nerdCopter lots of spare CPU cycles on that setup!

ctzsnooze commented:
Thanks @nerdCopter for your log.  The set point in angle mode is super smooth, looks great.  Hope it felt readily controlled in the air.
There is some gyro shake.  It's good to see that this doesn't end up in the angle mode setpoint (it could previously find its way in via accelerometer data).
![Screen Shot 2023-04-16 at 12 52 04](https://user-images.githubusercontent.com/11737748/232263673-bb089934-b862-4543-a895-e1a1cb127426.jpg)

TheIsotopes commented:
new testflight on `896c799` ... RX/TX Ghost Race250

[BTFL_BLACKBOX_LOG_Skinny_20230416.zip](https://github.com/betaflight/betaflight/files/11242245/BTFL_BLACKBOX_LOG_Skinny_20230416.zip)

TheIsotopes commented:
testflight with `a5e3d35` on Sbus

[BTFL_BLACKBOX_LOG_Mark1_20230417.zip](https://github.com/betaflight/betaflight/files/11250654/BTFL_BLACKBOX_LOG_Mark1_20230417.zip)

TheIsotopes commented:
another test flight with `a5e3d35` and on f405 everything looks fine again.
i think it was my bad to fly with 8k/8k dshot600, am now back on 8k/4k dshot 300.

[f405_f7_bbl.zip](https://github.com/betaflight/betaflight/files/11286926/f405_f7_bbl.zip)

imo its ready to merge.

ctzsnooze commented:
Setpoint and feedforward seem nicely smoothed, in that log, also.
![Screen Shot 2023-04-21 at 00 14 26](https://user-images.githubusercontent.com/11737748/233393838-ba19b413-3824-4f35-a1ad-e9699609cdb0.jpg)
Similarly the 'skinny' log shows a very clean RC link and very smooth set point, feedforward, and motors...
![Screen Shot 2023-04-21 at 00 17 05](https://user-images.githubusercontent.com/11737748/233394486-a99a1300-20a7-4edc-a09c-d3a884ef5813.jpg)

ctzsnooze commented:
The earlier F405 log has a peculiar issue with the FrSky radio link.  It seems like every fourth data packet is 'bigger' than the preceding three - like the Tx has a module sync issue.  This causes bumps in feedforward and P, leading to bumps in the motors.  I put a marker near two such steps; the regularity of the blips in the motors, and the coinciding large step in rcCommand, show what is going on.
The code is smoothing setpoint and feedforward correctly, I think, but these big steps seem 'real' to the FC, so they get through.
I don't see any issue with the motors, this problem is a straight up Tx issue I think.
Thanks again for testing!

![Screen Shot 2023-04-21 at 00 18 50](https://user-images.githubusercontent.com/11737748/233395628-4c8e85ca-fb8a-44f2-a909-a701e3947136.jpg)

TheIsotopes commented:
@ctzsnooze thank you for your diagnosis of the log files.

what would you call that flip at the beginning on screenshot? overshoot, undershoot or something else. 

With the f4 ultraskny, the flip is even worse at the beginning, so I wait a bit after min throttle and then perform a flip, so other flips look cleaner.

![Screenshot 2023-04-20 163458](https://user-images.githubusercontent.com/29702171/233401491-2762cf24-8be5-46af-8ecb-7cad04bdfebc.png)

KarateBrot commented on src/main/cli/cli.c:
```diff
@@ -4845,12 +4845,13 @@ static void cliRcSmoothing(const char *cmdName, char *cmdline)
                 cliPrintLine("NO SIGNAL");
             } else {
-                cliPrintLinef("%d.%03dms", avgRxFrameUs / 1000, avgRxFrameUs % 1000);
+                cliPrintLinef("%dHz", smoothedRxRateHz);
```
```suggestion
                cliPrintLinef("%.0fHz", rcSmoothingData->smoothedRxRateHz);
```

ctzsnooze replied:
Very cool, much better!  Didn't know about that method.

ctzsnooze replied:
Hmm @KarateBrot I get this error:
```
./src/main/cli/cli.c:4852:56: error: implicit conversion from 'float' to 'doubl' when passing argument to function [-Werror=double-promotion]
 4852 |                 cliPrintLinef("%.0fHz", rcSmoothingData->smoothedRxRateHz);
 ```

KarateBrot commented on src/main/cli/cli.c:
```diff
@@ -4845,12 +4845,13 @@ static void cliRcSmoothing(const char *cmdName, char *cmdline)
-            const uint16_t avgRxFrameUs = rcSmoothingData->averageFrameTimeUs;
-            cliPrint("# Detected RX frame rate: ");
-            if (avgRxFrameUs == 0) {
+            const uint16_t smoothedRxRateHz = lrintf(rcSmoothingData->smoothedRxRateHz);
+            int rxIntervalUs = getCurrentRxIntervalUs();
+            cliPrint("# Detected Rx frequency: ");
+            if (rxIntervalUs == 0) {
```
```suggestion
            cliPrint("# Detected Rx frequency: ");
            if (getCurrentRxIntervalUs() == 0) {
```

ctzsnooze replied:
OK, no worries, thanks.

KarateBrot commented on src/main/fc/parameter_names.h:
```diff
@@ -136,6 +134,7 @@
 #define PARAM_NAME_HORIZON_LIMIT_DEGREES "horizon_limit_degrees"
 #define PARAM_NAME_HORIZON_LIMIT_STICKS "horizon_limit_sticks"
 #define PARAM_NAME_HORIZON_IGNORE_STICKS "horizon_ignore_sticks"
+#define PARAM_NAME_HORIZON_DELAY "horizon_delay"
```
```suggestion
#define PARAM_NAME_HORIZON_DELAY_MS "horizon_delay_ms"
```

ctzsnooze replied:
Done

KarateBrot commented on src/main/fc/rc.c:
```diff
@@ -57,57 +57,54 @@
 static applyRatesFn *applyRates;
-static uint16_t currentRxRefreshRate;
+
+static uint16_t currentRxIntervalUs; // packet interval un microseconds
+static float currentRxRateHz; // packet interval un microseconds
```
```suggestion
static uint16_t currentRxIntervalUs; // packet interval in microseconds
static float currentRxRateHz;        // packet rate in hertz
```

ctzsnooze replied:
fixed, thanks

KarateBrot commented on src/main/flight/pid_init.c:
```diff
@@ -300,9 +286,13 @@ void pidInitConfig(const pidProfile_t *pidProfile)
+    pidRuntime.horizonLimitSticksInv = (pidProfile->pid[PID_LEVEL].D) ? 1.0f / pidRuntime.horizonLimitSticks : 1.0f;
     pidRuntime.horizonLimitDegrees = (float)pidProfile->horizon_limit_degrees;
+    pidRuntime.horizonLimitDegreesInv = (pidProfile->horizon_limit_degrees) ? 1.0f / pidRuntime.horizonLimitDegrees : 1.0f;
+    pidRuntime.horizonDelay = pidProfile->horizon_delay;
```
```suggestion
    pidRuntime.horizonDelayMs = pidProfile->horizon_delay_ms;
```

ctzsnooze replied:
Done

KarateBrot commented on src/main/flight/pid.c:
```diff
@@ -368,125 +346,87 @@ float pidApplyThrustLinearization(float motorOutput)
     // by compensating for the effect of yaw on roll while pitched, and on pitch while rolled
     float axisCoordination = pidRuntime.angleYawSetpoint;
-    if (pidRuntime.angleEarthRef) {
+    const float angleEarthRef = pidRuntime.angleEarthRef;
+    if (angleEarthRef) {
```
```suggestion
    if (pidRuntime.angleEarthRef) {
```

KarateBrot commented on src/main/flight/pid.c:
```diff
@@ -368,125 +346,87 @@ float pidApplyThrustLinearization(float motorOutput)
     // 0 when level, 90 when vertical, 180 when inverted (degrees):
-    float horizonLevelStrength = MAX((pidRuntime.horizonLimitDegrees - currentInclination) / pidRuntime.horizonLimitDegrees, 0.0f);
+
+    const float hlim = pidRuntime.horizonLimitDegrees;
+    const float hlimInv = pidRuntime.horizonLimitDegreesInv;
```
Not needed
```suggestion
```

KarateBrot commented on src/main/flight/pid.c:
```diff
@@ -368,125 +346,87 @@ float pidApplyThrustLinearization(float motorOutput)
-        const float horizonStickAttenuation = MAX((pidRuntime.horizonLimitSticks - absMaxStickDeflection) / pidRuntime.horizonLimitSticks, 0.0f);
+        // horizonIgnoreSticks:  0 = default; levelling attenuated by both attitude and sticks;
+        //                       1 = level attenuation only by attitude
+        const float hLimSticks = pidRuntime.horizonLimitSticks;
+        const float hLimSticksInv = pidRuntime.horizonLimitSticksInv;
+        const float horizGain = pidRuntime.horizonGain;
```
Not needed
```suggestion
```

ctzsnooze replied:
You're right, makes no difference.  Removed these declarations and optimised the code significantly.

KarateBrot commented on src/main/flight/pid.c:
```diff
@@ -368,125 +346,87 @@ float pidApplyThrustLinearization(float motorOutput)
+        const float horizGain = pidRuntime.horizonGain;
+        const float absMaxStickDeflection = MAX(fabsf(getRcDeflection(FD_ROLL)), fabsf(getRcDeflection(FD_PITCH)));
+        // 0-1, smoothed if RC smoothing is enabled
+        const float horizonStickAttenuation = MAX((hLimSticks - absMaxStickDeflection) * hLimSticksInv, 0.0f);
```
```suggestion
        const float horizonStickAttenuation = MAX((pidRuntime.horizonLimitSticks - absMaxStickDeflection) * pidRuntime.horizonLimitSticksInv, 0.0f);
```

ctzsnooze replied:
Done

KarateBrot commented on src/main/flight/pid.c:
```diff
@@ -981,44 +904,39 @@ void FAST_CODE pidController(const pidProfile_t *pidProfile, timeUs_t currentTim
+                    if (FLIGHT_MODE(HORIZON_MODE)) {
+                        maxAngleTargetAbs *= horizonLevelStrength; // reduce compensation while Horizon uses less levelling
+                    }
+                    float attenuateYawSetpoint = cos_approx(DEGREES_TO_RADIANS(maxAngleTargetAbs));
+                    currentPidSetpoint *= attenuateYawSetpoint;
```
```suggestion
                    currentPidSetpoint *= cos_approx(DEGREES_TO_RADIANS(maxAngleTargetAbs));  // Attenuate yaw setpoint
```

ctzsnooze replied:
I measured both options, they took exactly the same time - 86 cycles from before declaring maxAngleTargetAbs, to the debug.  Of those 86 cycles, 79 come from the cosApprox function.
I'll implement your suggestion.

haslinghuis commented on src/test/unit/pid_unittest.cc:
```diff
@@ -618,51 +596,51 @@ TEST(pidControllerTest, testCrashRecoveryMode)
+// 
+//     EXPECT_FLOAT_EQ(0, pidData[FD_ROLL].F);
+//     EXPECT_FLOAT_EQ(0, pidData[FD_PITCH].F);
+//     EXPECT_FLOAT_EQ(0, pidData[FD_YAW].F);
+// 
+//     // Match the stick to gyro to stop error
+//     setStickPosition(FD_ROLL, 1.0f);
+//     setStickPosition(FD_PITCH, -1.0f);
+//     setStickPosition(FD_YAW, -1.0f);
+// 
+//     pidController(pidProfile, currentTestTime());
+// 
+//     EXPECT_NEAR(2232.78, pidData[FD_ROLL].F, calculateTolerance(2232.78));
+//     EXPECT_NEAR(-2061.03, pidData[FD_PITCH].F, calculateTolerance(-2061.03));
+//     EXPECT_NEAR(-2061.03, pidData[FD_YAW].F, calculateTolerance(-2061.03));
+// 
+//     // Match the stick to gyro to stop error
+//     setStickPosition(FD_ROLL, 0.5f);
+//     setStickPosition(FD_PITCH, -0.5f);
+//     setStickPosition(FD_YAW, -0.5f);
+// 
+//     pidController(pidProfile, currentTestTime());
+// 
+//     EXPECT_NEAR(-558.20, pidData[FD_ROLL].F, calculateTolerance(-558.20));
+//     EXPECT_NEAR(515.26, pidData[FD_PITCH].F, calculateTolerance(515.26));
+//     EXPECT_NEAR(515.26, pidData[FD_YAW].F, calculateTolerance(515.26));
+// 
+//     setStickPosition(FD_ROLL, 0.0f);
+//     setStickPosition(FD_PITCH, 0.0f);
+//     setStickPosition(FD_YAW, 0.0f);
+// 
+//     for (int loop = 0; loop <= 15; loop++) {
+//         gyro.gyroADCf[FD_ROLL] += gyro.gyroADCf[FD_ROLL];
+//         pidController(pidProfile, currentTestTime());
+//     }
+// 
+//     EXPECT_FLOAT_EQ(0, pidData[FD_ROLL].F);
+//     EXPECT_FLOAT_EQ(0, pidData[FD_PITCH].F);
+//     EXPECT_FLOAT_EQ(0, pidData[FD_YAW].F);
+// }
```
Do we still need this?
```
// TEST(pidControllerTest, testFeedForward) // now in rc.c
// {
//     resetTest();
//     ENABLE_ARMING_FLAG(ARMED);
//     pidStabilisationState(PID_STABILISATION_ON);
// 
//     EXPECT_FLOAT_EQ(0, pidData[FD_ROLL].F);
//     EXPECT_FLOAT_EQ(0, pidData[FD_PITCH].F);
//     EXPECT_FLOAT_EQ(0, pidData[FD_YAW].F);
// 
//     // Match the stick to gyro to stop error
//     setStickPosition(FD_ROLL, 1.0f);
//     setStickPosition(FD_PITCH, -1.0f);
//     setStickPosition(FD_YAW, -1.0f);
// 
//     pidController(pidProfile, currentTestTime());
// 
//     EXPECT_NEAR(2232.78, pidData[FD_ROLL].F, calculateTolerance(2232.78));
//     EXPECT_NEAR(-2061.03, pidData[FD_PITCH].F, calculateTolerance(-2061.03));
//     EXPECT_NEAR(-2061.03, pidData[FD_YAW].F, calculateTolerance(-2061.03));
// 
//     // Match the stick to gyro to stop error
//     setStickPosition(FD_ROLL, 0.5f);
//     setStickPosition(FD_PITCH, -0.5f);
//     setStickPosition(FD_YAW, -0.5f);
// 
//     pidController(pidProfile, currentTestTime());
// 
//     EXPECT_NEAR(-558.20, pidData[FD_ROLL].F, calculateTolerance(-558.20));
//     EXPECT_NEAR(515.26, pidData[FD_PITCH].F, calculateTolerance(515.26));
//     EXPECT_NEAR(515.26, pidData[FD_YAW].F, calculateTolerance(515.26));
// 
//     setStickPosition(FD_ROLL, 0.0f);
//     setStickPosition(FD_PITCH, 0.0f);
//     setStickPosition(FD_YAW, 0.0f);
// 
//     for (int loop = 0; loop <= 15; loop++) {
//         gyro.gyroADCf[FD_ROLL] += gyro.gyroADCf[FD_ROLL];
//         pidController(pidProfile, currentTestTime());
//     }
// 
//     EXPECT_FLOAT_EQ(0, pidData[FD_ROLL].F);
//     EXPECT_FLOAT_EQ(0, pidData[FD_PITCH].F);
//     EXPECT_FLOAT_EQ(0, pidData[FD_YAW].F);
// }
```

ctzsnooze replied:
Thanks, Mark.  I left it there only because I have not been able to successfully port these unit tests across, now the feedforward code is in rc.c.  I simply cannot figure how to do it.  So these are left here as a 'reminder' of how the old tests were constructed.  Once we have the new unit tests in place, this can go.

borisbstyle replied:
Why are you removing tests? Those are exactly there to proove the flight behaviour is unchanged with refactoring.

haslinghuis replied:
Agree we need to rewrite the tests before merging

ctzsnooze replied:
I remove them from this place, because they cannot work in this location. 
I don't have a sufficient understanding of unit tests to implement the same tests, given that now that the code been moved new location.
I've asked many times for help with the unit tests.

ctzsnooze replied:
I checked these tests, they are somewhat useless.  They calculate feedforward from simulated setpointDelta values, and apply a transition value, but neither of those bits of maths actually check that the feedforward value is properly derived from changes in setpoint.

I've updated the pid_unittest so that it performs basically as before, in that it makes and artificial setpoint from stick values, computes a primitive setpointDelta, and then applies feedforward gain.  

We will need to create a proper feedforward unittest.  It probably will need to be in rc.c, where both the setpoint and feedforward code exists.  I've started work on that on [a local branch of mine](https://github.com/betaflight/betaflight/commit/5e94aeffac6c56ea21fd90ef4cddc24b9927d618)...

Currently the unit test on that branch fails in this manner, and I can't fix it.  If someone knows how, please share:-) :

```
compiling unit/rc_unittest.cc 
unit/rc_unittest.cc:157:5: error: use of undeclared identifier 'calculateFeedforward'
    calculateFeedforward(&pidRuntime, FD_ROLL); // **** FAILS !! *******
```

borisbstyle replied:
@ctzsnooze well not useless as you want to do tests as much as possible in isolation without too many external factors tested somewehere else. 
I haven't really looked too much in your code yet to understand it properly in how it can be fixed

ctzsnooze replied:
Thanks @borisbstyle - I have fixed the tests that can be done in the pid_unittest now.
Am working on proper tests of feedforward maths itself, in the rc.c side of things.
These were never done, previously.

haslinghuis commented on src/main/cli/settings.c:
```diff
@@ -1189,8 +1188,8 @@ const clivalue_t valueTable[] = {
-    { PARAM_NAME_FEEDFORWARD_BOOST,          VAR_UINT8 | PROFILE_VALUE,  .config.minmaxUnsigned = { 0, 50 }, PG_PID_PROFILE, offsetof(pidProfile_t, feedforward_boost) },
-    { PARAM_NAME_FEEDFORWARD_MAX_RATE_LIMIT, VAR_UINT8 | PROFILE_VALUE, .config.minmaxUnsigned = {0, 150}, PG_PID_PROFILE, offsetof(pidProfile_t, feedforward_max_rate_limit) },
+    { PARAM_NAME_FEEDFORWARD_BOOST,          VAR_UINT8 | PROFILE_VALUE, .config.minmaxUnsigned = { 0, 50 }, PG_PID_PROFILE, offsetof(pidProfile_t, feedforward_boost) },
+    { PARAM_NAME_FEEDFORWARD_MAX_RATE_LIMIT, VAR_UINT8 | PROFILE_VALUE, .config.minmaxUnsigned = {0, 200}, PG_PID_PROFILE, offsetof(pidProfile_t, feedforward_max_rate_limit) },
```
requires update in configurator ?

ctzsnooze replied:
Thanks, yes, not essential to change configurator, since I think very few people need above 150.

ctzsnooze commented on src/main/blackbox/blackbox.c:
```diff
@@ -1403,13 +1403,12 @@ static bool blackboxWriteSysinfo(void)
         BLACKBOX_PRINT_HEADER_LINE(PARAM_NAME_ANGLE_EARTH_REF, "%d",            currentPidProfile->angle_earth_ref);
-        BLACKBOX_PRINT_HEADER_LINE(PARAM_NAME_HORIZON_LIMIT_DEGREES, "%d",     currentPidProfile->horizon_limit_degrees);
+        BLACKBOX_PRINT_HEADER_LINE(PARAM_NAME_HORIZON_LIMIT_DEGREES, "%d",      currentPidProfile->horizon_limit_degrees);
+        BLACKBOX_PRINT_HEADER_LINE(PARAM_NAME_HORIZON_DELAY, "%d",              currentPidProfile->horizon_delay);
```
Good point.  The value is in ms * 10 unfortunately... so '50' means '500ms'.  
Should I still re-name to `horizon_delay_ms`?

ctzsnooze commented on src/main/fc/rc.h:
```diff
@@ -41,10 +41,12 @@ bool isMotorsReversed(void);
-float applyCurve(int axis, float deflection);
-bool getShouldUpdateFeedforward();
+
+#ifdef USE_FEEDFORWARD
+float getMaxRcRate(int axis);
+float getFeedforward(int axis);
+#endif
```
OK, I can do that for `getMaxRcRate` with no problem.
My concern with `getFeedforward` is that the function itself is within `#ifdef` for feedforward, so then we would have a declaration with no matching function if feedforward was not included.  So long as that is not a problem, then I'm good with both changes.

ctzsnooze commented on src/main/flight/pid.h:
```diff
@@ -236,6 +236,7 @@ typedef struct pidProfile_s {
     uint16_t tpa_breakpoint;                // Breakpoint where TPA is activated
     uint8_t angle_feedforward_smoothing_ms; // Smoothing factor for angle feedforward as time constant in milliseconds
     uint8_t angle_earth_ref;         // Control amount of "co-ordination" from yaw into roll while pitched forward in angle mode
+    uint8_t horizon_delay;           // delay when Horizon Strength increases, 50 = 500ms time constant
```
Done

ctzsnooze commented on src/main/flight/pid.h:
```diff
@@ -393,23 +389,26 @@ typedef struct pidRuntime_s {
     pt3Filter_t attitudeFilter[2];  // Only for ROLL and PITCH
-    pt3Filter_t angleFeedforwardPt3[XYZ_AXIS_COUNT];
+    pt1Filter_t horizonSmoothingPt1;
+    uint8_t horizonDelay;
```
Done

ctzsnooze commented on src/main/flight/pid_init.c:
```diff
@@ -239,11 +239,19 @@ void pidInitFilters(const pidProfile_t *pidProfile)
     const float k = pt3FilterGain(ATTITUDE_CUTOFF_HZ, pidRuntime.dT);
     const float angleCutoffHz = 1000.0f / (2.0f * M_PIf * pidProfile->angle_feedforward_smoothing_ms); // default of 80ms -> 2.0Hz, 160ms -> 1.0Hz, approximately
     const float k2 = pt3FilterGain(angleCutoffHz, pidRuntime.dT);
+    pidRuntime.horizonDelay = pidProfile->horizon_delay;
+    if (pidRuntime.horizonDelay) {
+        const float horizonSmoothingHz = 100.0f / (2.0f * M_PIf * pidProfile->horizon_delay); // default of 50 means 0.5s or 0.318Hz
```
Done

ctzsnooze commented on src/test/Makefile:
```diff
@@ -513,7 +513,7 @@ ifeq ($(shell $(CC) -v 2>&1 | grep -q "clang version" && echo "clang"),clang)
 # Travis reports CC_VERSION of 4.2.1
 CC_VERSION_MAJOR := $(firstword $(subst ., ,$(CC_VERSION)))
-CC_VERSION_CHECK_MIN := 7
+CC_VERSION_CHECK_MIN := 4
```
I'm still unsuccessful in porting the unit tests for feedforward after moving that code too rc.c

ctzsnooze commented on src/main/fc/rc.c:
```diff
@@ -57,57 +57,54 @@
-    }
-    return updateFf;
-}
+#define RX_INTERVAL_MIN_US     950   // 0.950ms to fit 1kHz without an issue
```
Done.

ctzsnooze commented on src/main/cli/settings.c:
```diff
@@ -1144,6 +1142,7 @@ const clivalue_t valueTable[] = {
     { PARAM_NAME_HORIZON_LIMIT_STICKS,   VAR_UINT8  | PROFILE_VALUE, .config.minmaxUnsigned = { 10, 200 }, PG_PID_PROFILE, offsetof(pidProfile_t, pid[PID_LEVEL].D) },
     { PARAM_NAME_HORIZON_LIMIT_DEGREES,  VAR_UINT8  | PROFILE_VALUE, .config.minmaxUnsigned = { 10,  250 }, PG_PID_PROFILE, offsetof(pidProfile_t, horizon_limit_degrees) },
     { PARAM_NAME_HORIZON_IGNORE_STICKS,  VAR_UINT8  | PROFILE_VALUE | MODE_LOOKUP, .config.lookup = { TABLE_OFF_ON }, PG_PID_PROFILE, offsetof(pidProfile_t, horizon_ignore_sticks) },
+    { PARAM_NAME_HORIZON_DELAY,          VAR_UINT8  | PROFILE_VALUE, .config.minmaxUnsigned = { 10,  250 }, PG_PID_PROFILE, offsetof(pidProfile_t, horizon_delay) }, 
```
Done, and range increased.

ctzsnooze commented on src/main/flight/pid.c:
```diff
@@ -228,6 +227,7 @@ void resetPidProfile(pidProfile_t *pidProfile)
         .tpa_breakpoint = 1350,
         .angle_feedforward_smoothing_ms = 80,
         .angle_earth_ref = 100,
+        .horizon_delay = 50, // 500ms time constant on any increase in horizon strength
```
Done

ctzsnooze commented on src/main/flight/pid.c:
```diff
@@ -368,125 +346,87 @@ float pidApplyThrustLinearization(float motorOutput)
-        horizonLevelStrength *= horizonStickAttenuation * pidRuntime.horizonGain;
+        horizonLevelStrength *= horizonStickAttenuation * horizGain;
+    }
+    if (pidRuntime.horizonDelay) {
```
Done

haslinghuis reviewed (changes requested):
Need to fix unit tests before merging

ctzsnooze commented:
Need someone to help get the 'container' for the unit tests.  I'll do the tests themselves, but I can't figure how to do them now the code is in a different place.  So if not one person here can help, then we should merge the PR without the unit tests.

ctzsnooze commented:
@borisbstyle @haslinghuis - in this PR:
- `getFeedforward(axis)` returns the feedforward value for the axis from rc.c, where it is calculated from incoming rc data.
- in pid.c, `pidSetpointDelta` is set to `getFeedforward(axis)`, or zero in Angle mode.
- then, simply:
```
        if (feedforwardGain > 0) {
            float feedForward = feedforwardGain * pidSetpointDelta;
            pidData[axis].F = feedForward;
        } else {
            pidData[axis].F = 0;
        }
```
In the old code, the functions `applyFeedforwardLimit`, `feedforwardInit`, `feedforwardApply`, and  `shouldApplyFeedforwardLimits` were included in the unit test.  They don't exist any more.
So I can appreciate that somehow the unit test needs to run `getFeedforward(axis)` and bring back values into pid.c.
I just don't know how to do that sort of stuff.

ctzsnooze commented on src/main/flight/pid.c:
```diff
@@ -368,125 +346,87 @@ float pidApplyThrustLinearization(float motorOutput)
-    angleFeedforward = pidRuntime.angleFeedforward[axis] * pidRuntime.angleFeedforwardGain;
-    // filter angle feedforward, heavily, at the PID loop rate, providing user control over time constant
+#ifdef USE_FEEDFORWARD
+    const float angleFeedforwardGain = pidRuntime.angleFeedforwardGain;
+    angleFeedforward = angleLimit * getFeedforward(axis) * angleFeedforwardGain * pidRuntime.maxRcRateInv[axis];
```
Done.

haslinghuis reviewed (approved):
@borisbstyle this should be good for now as we can improve the tests later

TheIsotopes commented:
@ctzsnooze here another testflight ... hope now that's the last before merge

[#12605_testflight.zip](https://github.com/betaflight/betaflight/files/11303702/12605_testflight.zip)

nerdCopter commented:
`b17cd84cf` LOS Horizon Mode.
`set debug_mode = RC_SMOOTHING_RATE`
ELRS; TX 500hz, 1:4 telem
[BTFL_BLACKBOX_LOG_APEX-6inch_20230423_123855_LOS_HORIZ_PR12605.BBL.zip](https://github.com/betaflight/betaflight/files/11304432/BTFL_BLACKBOX_LOG_APEX-6inch_20230423_123855_LOS_HORIZ_PR12605.BBL.zip)



```
# 

# version
# Betaflight / STM32F7X2 (S7X2) 4.5.0 Apr 23 2023 / 15:14:14 (b17cd84cf) MSP API: 1.46
# board: manufacturer_id: APEX, board_name: APEXF7

# rc_smoothing_info
# RC Smoothing Type: FILTER
# Detected Rx frequency: 503Hz
# Active setpoint cutoff: 25hz (manual)
# Active FF cutoff: 25hz (manual)
# Active throttle cutoff: 188hz (auto)

# status
MCU F722 Clock=216MHz, Vref=3.28V, Core temp=53degC
Stack size: 2048, Stack address: 0x20010000
Configuration: CONFIGURED, size: 3361, max available: 16384
Devices detected: SPI:1, I2C:1
Gyros detected: gyro 1 locked dma
GYRO=MPU6000, ACC=MPU6000, BARO=BMP280
OSD: MAX7456 (30 x 13)
BUILD KEY: 7e43be5093e31b54bc8383ab9a0594d0 (4.5.0-zulu)
System Uptime: 35 seconds, Current Time: 2023-04-23T17:41:19.109+00:00
CPU:34%, cycle time: 124, GYRO rate: 8064, RX rate: 500, System rate: 10
Voltage: 295 * 0.01V (0S battery - NOT PRESENT)
I2C Errors: 0
FLASH: JEDEC ID=0x00ef4018 16M
Arming disable flags: NOPREARM CLI MSP

# tasks
Task list             rate/hz  max/us  avg/us maxload avgload  total/ms   late    run reqd/us
00 - (         SYSTEM)     10       7       0    0.0%    0.0%         0      1    429       6
01 - (         SYSTEM)    995      13       0    1.2%    0.0%        23     72  42290       2
02 - (           GYRO)   8006      15       1   12.0%    1.4%       578      0 342169       0
03 - (         FILTER)   8006      29       9   23.2%    7.2%      3135      0 342169       0
04 - (            PID)   8006      64      28   51.2%   22.7%      9807      0 342169       0
05 - (            ACC)    992      18       2    1.7%    0.2%       131     45  42254       9
06 - (       ATTITUDE)    100      27      11    0.2%    0.1%        46      7   4694      22
07 - (             RX)    388      32      16    1.2%    0.6%       773     65  47997      18
08 - (         SERIAL)    100  296907       1 2969.0%    0.0%       559      0   4171      24
09 - (       DISPATCH)    993      14       0    1.3%    0.0%        27     61  42295       6
10 - (BATTERY_VOLTAGE)     50      15       3    0.0%    0.0%         4      3   2131      10
11 - (BATTERY_CURRENT)     50       7       0    0.0%    0.0%         1      6   2131       4
12 - ( BATTERY_ALERTS)      5       9       2    0.0%    0.0%         0      1    215       8
13 - (         BEEPER)    100      13       0    0.1%    0.0%         5      9   4170       9
14 - (           BARO)     40      41      23    0.1%    0.0%       119     10   5083      38
15 - (       ALTITUDE)    100      15       2    0.1%    0.0%        12      7   4170      11
16 - (      TELEMETRY)    499      19       0    0.9%    0.0%        36     21  20937       6
17 - (            OSD)     12      55      17    0.0%    0.0%       179     57  13227      32
19 - (            CMS)     20       9       2    0.0%    0.0%         1      3    852       8
20 - (        VTXCTRL)      5      35       1    0.0%    0.0%         3      4    215      34
21 - (        CAMCTRL)      5       6       0    0.0%    0.0%         0      0    215       5
23 - (    ADCINTERNAL)      1      11       3    0.0%    0.0%         0      2     43      10
24 - (       PINIOBOX)     20      15       2    0.0%    0.0%         2      2    852      11
25 - (SPEED_NEGOTIATION)    100      11       1    0.1%    0.0%         5      2   4170       5
RX Check Function                  14       0                        66
Total (excluding SERIAL)                                32.2%

# 
```

KarateBrot commented:
@haslinghuis Dismissing my review is not okay. I was about to take a final look over the PR before approving. Please never do this again.

haslinghuis commented:
@KarateBrot Just trying to keep the project going here. Assumed your request was fixed.

 We always can fix upon (any) work if it doesn't turn out. As this PR was hanging and multiple requests for merging were pending and tested were positive I deemed it ready for merge.

ctzsnooze commented:
Sorry @KarateBrot I thought that I had addressed your comments.  And I remain very grateful for the time you took to make those reviews.

My error was to mark those which I thought had been addressed, as 'resolved', which I did for housekeeping purposes, so I could track which had been dealt with.  I thought that this could always be reversed if we wanted to go back and check on how they were 'resolved'.

Unfortunately, I now see that, on merging, all review discussions marked as 'resolved' have vanished.  I didn't expect this.

It would be much better that all review discussions, including those marked as 'resolved', are retained on merging.
I don't know if this is a GitHub option.  In future I will leave all reviews 'open' even if they are 'resolved' because the discussions in those reviews can be useful in future.

The remaining review questions that I didn't resolve are still open, eg the query about the CLI print line function.

I wonder if those review discussions can be restored somehow?  Or if we can change our GitHub settings to not delete 'resolved' review discussions?

nerdCopter commented:
recommend author(s) set long-standing PR as draft until truly "ready".

nerdCopter commented:
we can still see some comments/unresolved via the "chaged files" tab.
e.g. https://github.com/betaflight/betaflight/pull/12605/files#r1172738403
